botocore/client.py

```diff
@@ -1091,13 +1092,34 @@ def _resolve_endpoint_ruleset(
                 request_context['auth_type'] = auth_type
                 if 'region' in signing_context and ignore_signing_region:
                     del signing_context['region']
+                self._maybe_warn_signing_region_mismatch(signing_context)
                 if 'signing' in request_context:
                     request_context['signing'].update(signing_context)
                 else:
                     request_context['signing'] = signing_context
 
         return endpoint_url, additional_headers
 
+    def _maybe_warn_signing_region_mismatch(self, signing_context):
+        legacy_signing_region = self._request_signer.region_name
+        signing_region = signing_context.get('region')
+        if (
+            signing_region is not None
+            and legacy_signing_region != signing_region
+            and not self._ruleset_resolver.uses_builtin_data_path
+            and self._ruleset_resolver.credential_scope_set
+        ):
```

Comment:
There's a variety of ways and places we can do this, but I think at the client level makes the most sense because we need state from the endpoint resolver and request signer. This also avoids adding a new attribute to the signer or adding additional data to the request context for a very niche case.

Reply:
I think we should probably pull this and do something closer to the example above. I'm also not clear on why we're associating the region name on the signer with `credentialScope` in the `endpoint.json` file. Won't this fire for any S3 bucket that's in a different region?

Reply:
No because S3 will never have credential scope configured on its ruleset. I'll still work on simplifying this though per your suggestion in the other comment.